src/data/CCDeque.ml

```diff
@@ -323,6 +327,107 @@ let to_rev_list q = fold (fun l x -> x::l) [] q
 
 let to_list q = List.rev (to_rev_list q)
 
+let size_cell_ = function
+  | Zero -> 0
+  | One _ -> 1
+  | Two _ -> 2
+  | Three _ -> 3
+
+(* filter over a cell *)
+let filter_cell_ f = function
+  | Zero -> Zero
+  | One x as c -> if f x then c else Zero
+  | Two (x,y) as c ->
+    let fx = f x in
+    let fy = f y in
+    begin match fx, fy with
+      | true, true -> c
+      | true, false -> One x
+      | false, true -> One y
+      | _ -> Zero
+    end
+  | Three (x,y,z) as c ->
+    let fx = f x in
+    let fy = f y in
+    let fz = f z in
+    begin match fx, fy, fz with
+      | true, true, true -> c
+      | true, true, false -> Two (x,y)
+      | true, false, true -> Two (x,z)
+      | true, false, false -> One x
+      | false, true, true -> Two (y,z)
+      | false, true, false -> One y
+      | false, false, true -> One z
+      | false, false, false -> Zero
+    end
+
+let filter_in_place (d:_ t) f : unit =
+  (* update size, compute new cell *)
+  let update_local_ n =
+    d.size <- d.size - size_cell_ n.cell;
+    let new_cell = filter_cell_ f n.cell in
+    d.size <- d.size + size_cell_ new_cell;
+    new_cell
+  in
+  let rec loop ~stop_at n : unit =
+    let n_prev = n.prev in
+    let n_next = n.next in
+    if n != stop_at then (
+      let new_cell = update_local_ n in
+      (* merge into previous cell *)
+      begin match n_prev.cell, new_cell with
+        | _, Zero -> remove_node_ n
+        | Zero, _ -> remove_node_ n; n_prev.cell <- new_cell;
+        | Three _, _ -> n.cell <- new_cell
+        | One x, One y -> remove_node_ n; n_prev.cell <- Two (x,y)
+        | One (x), Two (y,z)
+        | Two (x,y), One z -> remove_node_ n; n_prev.cell <- Three (x,y,z)
+        | One x, Three (y,z,w)
+        | Two (x,y), Two (z,w) -> n_prev.cell <- Three (x,y,z); n.cell <- One w
+        | Two (x,y), Three (z,w1,w2) -> n_prev.cell <- Three (x,y,z); n.cell <- Two (w1,w2)
+      end;
+      loop ~stop_at n_next;
+    );
+  in
+  d.cur.cell <- update_local_ d.cur; (* special case for first cell *)
+  loop ~stop_at:d.cur d.cur.next
```

Comment:
The side effect is that the first node will be a Three. But does not seem to be a problem.